### ISSUE TITLE
Debugging Stratum Mining Configuration

### ISSUE THREAD

**Issue by stone212** [paritytech/parity-ethereum#10028]
EDIT: Re-posting original issue

I can not mine on a private blockchain using parity in stratum mode and ethminer.

Parity Ethereum version:` 2.1.9`
ethminer:` 0.16`
Operating system: `Ubuntu 18.04`
GPU: 
```
01:00.0 VGA compatible controller: NVIDIA Corporation GP104 (rev a1)
03:00.0 VGA compatible controller: NVIDIA Corporation GP104 (rev a1)
```

Here is the output from ethminer:

```
m 01:12:05 ethminer ethminer 0.16.1
 m 01:12:05 ethminer Build: linux/release
cu 01:12:05 ethminer Using grid size: 8,192, block size: 128
 i 01:12:05 ethminer Configured pool 127.0.0.1:8008
 i 01:12:05 main     Selected pool 127.0.0.1:8008
 i 01:12:05 stratum  Stratum mode detected: STRATUM
 i 01:12:05 stratum  Subscribed!
 i 01:12:05 stratum  Authorized worker 
 i 01:12:05 stratum  Established connection with 127.0.0.1:8008 at  [127.0.0.1:8008]
 i 01:12:05 stratum  Spinning up miners...
 i 01:12:05 cuda-0   No work. Pause for 3 s.
 i 01:12:05 cuda-1   No work. Pause for 3 s.
 i 01:12:08 cuda-0   No work. Pause for 3 s.
 i 01:12:08 cuda-1   No work. Pause for 3 s.
 m 01:12:10 ethminer Speed 0.00 Mh/s gpu0 0.00 gpu1 0.00 [A0] Time: 00:00
 i 01:12:11 cuda-1   No work. Pause for 3 s.
 i 01:12:11 cuda-0   No work. Pause for 3 s.
 i 01:12:14 cuda-1   No work. Pause for 3 s.
 i 01:12:14 cuda-0   No work. Pause for 3 s.
```

Here is how I run ethminer:

`/opt/ethminer/bin/ethminer -R -U -P stratum+tcp://127.0.0.1:8008`

Here is my parity .toml file:

```
[parity]
chain = "/home/parity/my-chain-spec.json"

[network]
reserved_peers = "/home/parity/reserved-peers.txt"
min_peers = 15

[rpc]
hosts = [""]
cors = [""]

[account]
unlock = ["0x00..."]
password = ["/home/parity/keyPass.txt"]

[mining]
author = "0x00..."

[stratum]
```

`netstat -tulpn | grep '8008'` shows:

```
tcp        0      0 127.0.0.1:8008          0.0.0.0:*               LISTEN      6699/parity
```

Do you see something I do not see?

ethminer works when I use it on this worker with the same flags connected to a stratum pool.

Parity has 15 peers (and that is min_peers in toml file).

Thank you.

**Comment by tomusdrw** [comment]
@stone212 could you check if `--force-sealing` resolves the issue? Are there any transactions in the network or you want to mine empty blocks? Are there any blocks in the network already or do you start from genesis?

**Comment by stone212** [comment]
@tomusdrw 
`--force-sealig` does not solve the issue, just tried now

There are no transactions in the network at this moment so of course some time I will want to mine full blocks but now only empty blocks.

The chain height is in the thousands.  Most of those are empty blocks.

**Comment by tomusdrw** [comment]
@stone212 I think it still might be a valid issue, could you run with `-lminer=trace,stratum=trace`?

**Comment by stone212** [comment]
I see things like:
```
2018-12-07 02:26:34  event.loop TRACE stratum  Subscription request from V4(127.0.0.1:42510)
```
.
.
.
```
2018-12-07 02:26:34  event.loop TRACE miner  prepare_pending_block: Miner received request (was 0, now 1357639) - waking up.
2018-12-07 02:26:34  event.loop TRACE stratum  New worker # registered
2018-12-07 02:27:00  IO Worker #3 INFO import    15/15 peers   78 KiB chain 21 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-12-07 02:27:30  IO Worker #2 INFO import    15/15 peers   78 KiB chain 21 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
```
.
.
.
```
2018-12-07 02:29:05  IO Worker #2 INFO import    15/15 peers   78 KiB chain 21 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-12-07 02:29:34  event.loop TRACE stratum  Subscription request from V4(127.0.0.1:53546)
2018-12-07 02:29:34  event.loop TRACE miner  prepare_pending_block: entering
2018-12-07 02:29:34  event.loop TRACE miner  prepare_pending_block: have_work=true
2018-12-07 02:29:34  event.loop TRACE stratum  New worker # registered
2018-12-07 02:29:40  IO Worker #1 INFO import    15/15 peers   78 KiB chain 21 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,
```

That's where we are now.  I'm monitoring.

**Comment by stone212** [comment]
It basically repeats that last chunk.

**Comment by stone212** [comment]
@tomusdrw 

I can mine using ethminer 0.12 on the same machine.  The command I use is:

`./ethminer-0.12/bin/ethminer -G -SP 127.0.0.1:8008`

That does not mean the problem is with ethminer 0.16.  But I will post an issue there too.

**Comment by stone212** [comment]
@tomusdrw 

In the ethminer thread @AndreaLanfranchi says this and I agree that it seems most likely:

> if connection works with stratum+tcp without the 1 and you keep getting no-work the only reason is that parity node is not pushing jobs.

This is what the original error appears to be.  Here are a few questions that might help:

1. What stratum mode does Parity 2.1.9 default to?

2. Does Parity 2.1.9 offer `-eth-proxy` compatibility mode? (when I tried ethminer 0.16 with `-eth-proxy` compatibility I got a different error: 
```
 i 23:53:08 stratum  Stratum mode detected: ETHPROXY Compatible
 i 23:53:08 stratum  Could not login:code:-32601 message:Method not found 
```

Entire thread with ethminer: https://github.com/ethereum-mining/ethminer/issues/1763#issuecomment-445439785

**Comment by stone212** [comment]
@tomusdrw Upgrading to ethminer 0.18(alpha) got me mining again.

But I think the questions about Parity stratum modes still could be important.  I will leave the issue open to help you debug if you want.

**Comment by joshua-mir** [comment]
closing as it doesn't seem like this was an issue with parity

parity-ethereum doesn't support ethproxy mode (hint is in the name, the mode is for proxy software between an ethereum node's eth_getWork rpc method and miners) nor does it support stratum modes other than `stratum+tcp` ("standard stratum" which most pools do not use because they use things like authentication and share accounting which often requires a proxy between their nodes and miners)

**Comment by Parean** [comment]
@joshua-mir I have the same issue and I don't think that the matter is in the ethminer, because the same version of the miner works fine if I connect to the ropsten testnet. For some reason, when working with a private network, the parity sends a job only once, in response to mining.subscribe request. But the miner ignores this first job because it waits for the job exactly in the notification, and not in the response to the request. Even if this is the expected behavior, it is not clear why the behavior of parity is different in the private network and ropsten. At first I thought that the node will send a new job if I generate a new transaction and change block template. I tried running parity with the options "--reseal-on-txs=all --force-sealing" and call eth_sendRawTransaction several times, but it didn't help.  How can I make the parity repeatedly send jobs to a miner in a private network?

**Comment by Parean** [comment]
Also, I would be grateful if someone would tell me if there is a different, simpler method of generating new blocks in a private network at the moment. Without using ethminer.

**Comment by joshua-mir** [comment]
@Parean it might be worthwhile looking into the logs generated when running with `--logging miner`

Also, unless you are testing things related specifically to PoW networks, I would recommend using a PoA consensus algorithm (Aura/AuthorityRound for Parity-only networks, Clique if you want to connect with geth nodes as well) for private network - there is no need to run external mining software in either of those scenarios.

**Comment by Parean** [comment]
Yes, I know. But, unfortunately, I need to test the process of mining in the PoW network.

Here are the logs of the node running with the parameters `--logging engine, miner`. I called `eth_sendRawTransaction`   method three times while it was running.

```

2019-07-30 14:18:00  main INFO parity_ethereum::run  Starting Parity-Ethereum/v2.4.6-stable-c6f5bc443-20190522/x86_64-linux-gnu/rustc1.31.1
2019-07-30 14:18:00  main INFO parity_ethereum::run  Keys path /home/denis/.local/share/io.parity.ethereum/keys/TEST_CHAIN
2019-07-30 14:18:00  main INFO parity_ethereum::run  DB path /home/denis/.local/share/io.parity.ethereum/chains/TEST_CHAIN/db/5d4d6db40ae79a04
2019-07-30 14:18:00  main INFO parity_ethereum::run  State DB configuration: fast
2019-07-30 14:18:00  main INFO parity_ethereum::run  Operating mode: active
2019-07-30 14:18:01  main INFO ethcore_service::service  Configured for TEST_CHAIN using Ethash engine
2019-07-30 14:18:01  main TRACE miner  minimal_gas_price: recalibrating...
2019-07-30 14:18:01  main DEBUG miner  minimal_gas_price: Got gas price! 0
2019-07-30 14:18:01  main TRACE miner  prepare_pending_block: entering
2019-07-30 14:18:01  main TRACE miner  prepare_pending_block: have_work=false
2019-07-30 14:18:01  main TRACE miner  prepare_block: No existing work - making new block
2019-07-30 14:18:01  main DEBUG miner  Attempting to push 0 transactions.
2019-07-30 14:18:01  main DEBUG miner  Pushed 0 transactions in 0 ms
2019-07-30 14:18:01  main TRACE miner  prepare_work: Checking whether we need to reseal: orig=None last=None, this=0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91
2019-07-30 14:18:01  main TRACE miner  prepare_work: Pushing a new, refreshed or borrowed pending 0x5884…1c91...
2019-07-30 14:18:01  main TRACE miner  prepare_work: leaving (last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91))
2019-07-30 14:18:01  main TRACE miner  prepare_pending_block: Miner received request (was 0, now 0) - waking up.
2019-07-30 14:18:01  main TRACE miner  prepare_pending_block: entering
2019-07-30 14:18:01  main TRACE miner  prepare_pending_block: have_work=true
2019-07-30 14:18:01  main TRACE miner  update_sealing
2019-07-30 14:18:01  main TRACE miner  requires_reseal: sealing enabled
2019-07-30 14:18:01  main TRACE miner  requires_reseal: should_disable_sealing=false; forced=true, has_local=true, internal=None, had_requests=true
2019-07-30 14:18:01  main TRACE miner  update_sealing: preparing a block
2019-07-30 14:18:01  main TRACE miner  prepare_block: Already have previous work; updating and returning
2019-07-30 14:18:01  main DEBUG miner  Attempting to push 0 transactions.
2019-07-30 14:18:01  main DEBUG miner  Pushed 0 transactions in 0 ms
2019-07-30 14:18:01  main TRACE miner  update_sealing: engine does not seal internally, preparing work
2019-07-30 14:18:01  main TRACE miner  prepare_work: Checking whether we need to reseal: orig=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91) last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91), this=0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91
2019-07-30 14:18:01  main TRACE miner  prepare_work: leaving (last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91))
2019-07-30 14:18:01  main INFO parity_ws  Listening for new connections on 127.0.0.1:8547.
2019-07-30 14:18:06  IO Worker #2 INFO network  Public node URL: enode://5940d8345fbf3cbc4a7d27d8c4de60b7fed0d321d38570dd8cab696a59815249d42719ca23bf44dff58675c410b6a028f06171a930127d73d96aa188d7e95cf8@{ip}:30321
2019-07-30 14:18:10  jsonrpc-eventloop-0 TRACE miner  prepare_pending_block: entering
2019-07-30 14:18:10  jsonrpc-eventloop-0 TRACE miner  prepare_pending_block: have_work=true
2019-07-30 14:18:31  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,    0 µs
2019-07-30 14:19:01  IO Worker #1 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,    0 µs
2019-07-30 14:19:31  IO Worker #2 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,    0 µs
2019-07-30 14:20:01  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,    0 µs
2019-07-30 14:20:31  IO Worker #1 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,    0 µs
2019-07-30 14:21:01  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  731 µs
2019-07-30 14:21:31  IO Worker #2 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  731 µs
2019-07-30 14:22:01  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  731 µs
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  prepare_pending_block: entering
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  prepare_pending_block: have_work=true
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  update_sealing
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  requires_reseal: sealing enabled
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  requires_reseal: should_disable_sealing=false; forced=true, has_local=true, internal=None, had_requests=true
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  update_sealing: preparing a block
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  prepare_block: Already have previous work; updating and returning
2019-07-30 14:22:08  jsonrpc-eventloop-0 DEBUG miner  Attempting to push 0 transactions.
2019-07-30 14:22:08  jsonrpc-eventloop-0 DEBUG miner  Pushed 0 transactions in 0 ms
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  update_sealing: engine does not seal internally, preparing work
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  prepare_work: Checking whether we need to reseal: orig=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91) last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91), this=0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91
2019-07-30 14:22:08  jsonrpc-eventloop-0 TRACE miner  prepare_work: leaving (last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91))
2019-07-30 14:22:31  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  596 µs
2019-07-30 14:23:01  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  596 µs
2019-07-30 14:23:31  IO Worker #1 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  466 µs
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  prepare_pending_block: entering
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  prepare_pending_block: have_work=true
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  update_sealing
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  requires_reseal: sealing enabled
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  requires_reseal: should_disable_sealing=false; forced=true, has_local=true, internal=None, had_requests=true
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  update_sealing: preparing a block
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  prepare_block: Already have previous work; updating and returning
2019-07-30 14:23:50  jsonrpc-eventloop-0 DEBUG miner  Attempting to push 0 transactions.
2019-07-30 14:23:50  jsonrpc-eventloop-0 DEBUG miner  Pushed 0 transactions in 0 ms
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  update_sealing: engine does not seal internally, preparing work
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  prepare_work: Checking whether we need to reseal: orig=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91) last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91), this=0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91
2019-07-30 14:23:50  jsonrpc-eventloop-0 TRACE miner  prepare_work: leaving (last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91))
2019-07-30 14:24:01  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  466 µs
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  prepare_pending_block: entering
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  prepare_pending_block: have_work=true
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  update_sealing
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  requires_reseal: sealing enabled
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  requires_reseal: should_disable_sealing=false; forced=true, has_local=true, internal=None, had_requests=true
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  update_sealing: preparing a block
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  prepare_block: Already have previous work; updating and returning
2019-07-30 14:24:27  jsonrpc-eventloop-2 DEBUG miner  Attempting to push 0 transactions.
2019-07-30 14:24:27  jsonrpc-eventloop-2 DEBUG miner  Pushed 0 transactions in 0 ms
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  update_sealing: engine does not seal internally, preparing work
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  prepare_work: Checking whether we need to reseal: orig=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91) last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91), this=0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91
2019-07-30 14:24:27  jsonrpc-eventloop-2 TRACE miner  prepare_work: leaving (last=Some(0x5884b5a1242231456e5980593048269067b0509bb3db5f9d6c2dbd698f8e1c91))
2019-07-30 14:24:31  IO Worker #0 INFO import     0/25 peers   8 KiB chain 5 KiB db 0 bytes queue 448 bytes sync  RPC:  0 conn,    0 req/s,  466 µs


```

**Comment by joshua-mir** [comment]
@Parean actually, could you make a new issue about this? This actually looks `txqueue` related, as the transactions never actually appear to have gotten to the miner.
